### PR TITLE
(Closes #86) Add omp offload option to tracer advection

### LINKED ISSUE #86: Add OpenMP offload target for the tracer advection benchmark
We want to demonstrate OpenMP offload for the PSyclone paper and thus it makes sense to add it here. I'll work off `PSyclone/examples/nemo/eg1` for this.

### PR DESCRIPTION
**Pull request by arporter**

### CODE REVIEW

**Comment by arporter** [comment]
Ready for review now. As suggested in #86, I've just copied in the script from PSyclone/examples/nemo/eg1 for now. This will need updating once the new transformations are on master.

**Comment by arporter** [comment]
Performance of this version is currently not very good because some loops are left on the CPU due to complex loop bounds (involving UBOUND/LBOUND).

**Comment by sergisiso** [comment]
I was kind of expecting this I will review it this afternoon and when https://github.com/stfc/PSyclone/pull/1790 is merged this can be fixed. Looking at the source I think it should be feasible to bring everything to the GPU as it looks very much like parts on NEMO that I know are offloaded.

File: benchmarks/nemo/tracer_advection/original/Makefile [hunk 1]
```diff
@@ -46,14 +50,24 @@ tra_adv_omp: ../scripts/omp_trans.py
 	${MAKE} PROF_LIB_INC="-I../${DL_TIMER_DIR}/src" \
             LDFLAGS="${LDFLAGS} ../${DL_TIMER_DIR}/${DL_TIMER_NAME}" -C $@
 
+tra_adv_omp_gpu: ../scripts/omp_gpu_trans.py
```
Comment (sergisiso) [inline benchmarks/nemo/tracer_advection/original/Makefile:53]: Shouldn't `./tra_adv.F90` be also one of the rule prerequesites in case it is modified it needs to be re-psycloned and recompiled?
Also in the other targets

File: benchmarks/nemo/tracer_advection/original/Makefile [hunk 2]
```diff
@@ -46,14 +50,24 @@ tra_adv_omp: ../scripts/omp_trans.py
...
+	mkdir -p $@
+	${PSYCLONE} -s ../scripts/omp_gpu_trans.py -opsy $@/tra_adv.f90 ./tra_adv.F90
+	cp Makefile_gen $@/Makefile
+	${MAKE} PROF_LIB_INC="-I../${DL_TIMER_DIR}/src" \
+            LDFLAGS="${LDFLAGS} ${OMPTARGETFLAGS} ../${DL_TIMER_DIR}/${DL_TIMER_NAME}" -C $@
```
Comment (sergisiso) [inline benchmarks/nemo/tracer_advection/original/Makefile:58]: I would expect:
```
source <compiler_setup>/nvidia.sh
make tra_adv_omp_gpu
```
to work out of the box. But it doesn't because it requires the manage memory flag and manually update some environment variable.

We could add a new envvar in nvidia.sh
`export OMPTARGET_UM="-gpu=managed"`
(UM as unified memory because its a more standard term than managed memory across compilers)

and then add:
```
     cp Makefile_gen $@/Makefile
ifndef OMPTARGET_UM 
    $(error This implementation requires OpenMP unified memory, but the OMPTARGET_UM environment variable is undefined)
endif
     ${MAKE} PROF_LIB_INC="-I../${DL_TIMER_DIR}/src" \
```
This would allow it to work for nvidia directly but also provide and useful error message for other compilers instead of letting them compile and fail at runtime.

File: benchmarks/nemo/tracer_advection/original/Makefile (hunk 1 above)
Comment (sergisiso) [reply benchmarks/nemo/tracer_advection/original/Makefile:53]: (Although they are .PHONY - so they are recompiled no matter what)

**Review by sergisiso — CHANGES_REQUESTED** [review]
@arporter See inline comments. Also if I try to run it I get:
```
$ JPI=130 JPJ=130 JPK=130 IT=100 ./tra_adv_omp_gpu/tra_adv.exe
Checksum for domain  130 x 130 x 130 ( 100 iterations) =  0.0000000000000000E+00
```
with the checksum being 0 instead of the right number. Is that what you get?

**Comment by arporter** [comment]
I hadn't noticed the checksum problem. It turns out that OMPTargetTrans doesn't exclude CodeBlock nodes (and it should) so the final loop that computes the checksum gets put in an `omp target` block. Quite what the compiler does with that I'm not sure. For now I've extended the transformation script to exclude loops that contain CodeBlocks and now I get a non-zero checksum. I'll make a PSyclone issue for this.

File: benchmarks/nemo/tracer_advection/original/README.md [hunk 3]
```diff
@@ -20,13 +29,20 @@ environment variables:
...
 
     export F90=gfortran
     export F90FLAGS=-O3
     (export OMPFLAGS=-fopenmp)
     export MPIF90=mpif90
 
+To use the NVIDIA compiler, OpenMP offload and managed memory:
+
+    export F90=nvfortran
+    export F90FLAGS="-O3 -gpu=managed"
```
[the remark this answers is not in the document]
Comment (arporter) [reply benchmarks/nemo/tracer_advection/original/README.md:42]: Yes, done (and in nvidia.sh too).

File: benchmarks/nemo/tracer_advection/original/Makefile (hunk 1 above)
Comment (arporter) [reply benchmarks/nemo/tracer_advection/original/Makefile:53]: Yes, you are right - done for all the targets.

File: benchmarks/nemo/tracer_advection/original/Makefile (hunk 2 above)
Comment (arporter) [reply benchmarks/nemo/tracer_advection/original/Makefile:64]: That's a good idea and I've done it. I've also altered the way the various variables are passed down to the `Makefile_gen` to make it a bit clearer what's going on.

File: benchmarks/nemo/tracer_advection/original/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (arporter) [reply benchmarks/nemo/tracer_advection/original/README.md]: Actually, it's now in `UMEMFLAGS`, as per your suggestion below.

**Comment by arporter** [comment]
Checksum now matches for serial and OMP offload versions. Ready for another look.

File: benchmarks/nemo/tracer_advection/scripts/omp_gpu_trans.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            # nodes.
+            if loop.loop_type == "levels" and not loop.walk(CodeBlock):
+                try:
+                    omp_target_trans.apply(loop)
+
+                    omp_loop_trans.apply(loop)
+                except TransformationError:
+                    # This loop can not be transformed, proceed to next loop
+                    continue
+
```
Comment (sergisiso) [inline benchmarks/nemo/tracer_advection/scripts/omp_gpu_trans.py:83]: Since we are here, to make it fast I went away with the loop_type and rely on the dependency analysis trying to paralelise each loop starting from the outer loop. I this code all outer loops but `do jt = 1, itn_count, 1` should succeed. Can you try if replacing this bit with the following gives better performance:
```
  # TODO PSyclone/#1815: OMPTargetTrans does not exclude CodeBlock
  # nodes.
  if loop.walk(CodeBlock):
      continue

  if loop.ancestor(Directive):
      # Skip if an outer loop is already parallelised
      continue

  try:
      omp_loop_trans.apply(loop)
      # Only add the target directive if the OMPLoop was successfully applied.
      omp_target_trans.apply(loop.parent.parent)
  except TransformationError as err:
      # This loop can not be transformed, proceed to next loop
      print("Loop not parallelised because:", str(err))
      continue
```

File: benchmarks/nemo/tracer_advection/original/Makefile [hunk 5]
```diff
@@ -21,39 +27,53 @@ DL_TIMER_NAME = libdl_timer_omp.a
...
 	${MAKE} -C tra_adv_serial $@
 	${MAKE} -C tra_adv_acc $@
 	${MAKE} -C tra_adv_omp $@
+	${MAKE} -C tra_adv_omp_gpu $@
 
 allclean:
 	rm -rf tra_adv_serial
 	rm -rf tra_adv_acc
+	rm -rf tra_adv_omp
+	rm -rf tra_adv_omp_gpu
```
Comment (sergisiso) [inline benchmarks/nemo/tracer_advection/original/Makefile:79]: allclean didn't work from me because an old mod file in dl_timer, I think you need to add:
`${MAKE} -C ${DL_TIMER_DIR} allclean`

It not part of this PR but the problem is also present in the other folders of tracer_advection

File: benchmarks/nemo/tracer_advection/original/Makefile [hunk 6]
```diff
@@ -21,39 +27,53 @@ DL_TIMER_NAME = libdl_timer_omp.a
...
+	${PSYCLONE} -s ../scripts/omp_gpu_trans.py -opsy $@/tra_adv.f90 ./tra_adv.F90
+	cp Makefile_gen $@/Makefile
+	${MAKE} PROF_LIB_INC="-I../${DL_TIMER_DIR}/src" \
+            FORT_FLAGS="${F90FLAGS} ${OMPTARGETFLAGS} ${UMEMFLAGS} -I../${DL_TIMER_DIR}/src" \
+            LDFLAGS="${LDFLAGS} ${OMPTARGETFLAGS} ../${DL_TIMER_DIR}/${DL_TIMER_NAME}" -C $@
+
 dl_timer:
 	make -C ${DL_TIMER_DIR} sm_lib
 
 clean:
```
Comment (sergisiso) [inline benchmarks/nemo/tracer_advection/original/Makefile:69]: clean does not work if one of the target folders is not created, for instance `${MAKE} -C tra_adv_serial $@` fails if tra_adv_serial does not exist and the rest are not executed. You may need to check if the directory exist or use bash "or" commands among each make.

**Review by sergisiso — CHANGES_REQUESTED** [review]
Almost there. I had some issues trying to execute the make clean commands. And I also suggested the change that will make the OMP Target faster if you want to include it in this PR.

File: benchmarks/nemo/tracer_advection/scripts/omp_gpu_trans.py (hunk 4 above)
Comment (sergisiso) [reply benchmarks/nemo/tracer_advection/scripts/omp_gpu_trans.py:83]: Adding:
```
hoist_trans = HoistTrans()
for statement in invoke.schedule.walk(Statement):
    try:
        hoist_trans.apply(statement)
    except TransformationError as err:
        pass
```
before the OpenMP insertion loop may also help. (unfortunately repeated scalar assingments with the same value are not hoisted which would help here, I will open a ticket for this)

File: benchmarks/nemo/tracer_advection/original/Makefile (hunk 5 above)
Comment (arporter) [reply benchmarks/nemo/tracer_advection/original/Makefile:79]: Done.

File: benchmarks/nemo/tracer_advection/original/Makefile (hunk 6 above)
Comment (arporter) [reply benchmarks/nemo/tracer_advection/original/Makefile:69]: Oh yes, fixed.

File: benchmarks/nemo/tracer_advection/scripts/omp_gpu_trans.py (hunk 4 above)
Comment (arporter) [reply benchmarks/nemo/tracer_advection/scripts/omp_gpu_trans.py:83]: I tried the first part of this but hit #1787 when I tried to collapse the loops. Without the collapsing the performance is no better.

**Comment by arporter** [comment]
Ready for another look now.

**Review by sergisiso — APPROVED** [review]
@arporter All suggestions have been addressed. I tested that it compiles and produces the expected results on the Glados system. This is approved for merging.
